services/antivirus/pkg/service/service.go

```diff
@@ -0,0 +1,205 @@
+		if err != nil {
+			return av, err
+		}
+
+		av.m = b.Bytes()
+	}
+
+	return av, nil
+}
+
+// Antivirus defines implements the business logic for Service.
+type Antivirus struct {
+	c *config.Config
+	l log.Logger
+	s Scanner
+	o events.PostprocessingOutcome
+	m uint64
+
+	client *http.Client
+}
+
+// Run runs the service
+func (av Antivirus) Run() error {
+	evtsCfg := av.c.Events
+
+	var rootCAPool *x509.CertPool
+	if evtsCfg.TLSRootCACertificate != "" {
+		rootCrtFile, err := os.Open(evtsCfg.TLSRootCACertificate)
+		if err != nil {
+			return err
+		}
+
+		var certBytes bytes.Buffer
+		if _, err := io.Copy(&certBytes, rootCrtFile); err != nil {
+			return err
+		}
+
+		rootCAPool = x509.NewCertPool()
+		rootCAPool.AppendCertsFromPEM(certBytes.Bytes())
+		evtsCfg.TLSInsecure = false
+	}
+
+	stream, err := stream.NatsFromConfig(stream.NatsConfig(av.c.Events))
+	if err != nil {
+		return err
+	}
+
+	ch, err := events.Consume(stream, "antivirus", events.StartPostprocessingStep{})
+	if err != nil {
+		return err
+	}
+
+	for e := range ch {
+		ev := e.Event.(events.StartPostprocessingStep)
+		if ev.StepToStart != events.PPStepAntivirus {
+			continue
+		}
+
+		var errmsg string
+		res, err := av.process(ev)
+		if err != nil {
+			errmsg = err.Error()
+		}
+
+		outcome := events.PPOutcomeContinue
+		if res.Infected {
+			outcome = av.o
+		}
+
+		av.l.Info().Str("uploadid", ev.UploadID).Interface("resourceID", ev.ResourceID).Str("virus", res.Description).Str("outcome", string(outcome)).Str("filename", ev.Filename).Str("user", ev.ExecutingUser.GetId().GetOpaqueId()).Bool("infected", res.Infected).Msg("File scanned")
+		if err := events.Publish(stream, events.PostprocessingStepFinished{
+			Outcome:       outcome,
+			UploadID:      ev.UploadID,
+			ExecutingUser: ev.ExecutingUser,
+			Filename:      ev.Filename,
+			Result: events.VirusscanResult{
+				Infected:    res.Infected,
+				Description: res.Description,
+				Scandate:    time.Now(),
+				ResourceID:  ev.ResourceID,
+				ErrorMsg:    errmsg,
+			},
+		}); err != nil {
+			av.l.Fatal().Err(err).Str("uploadid", ev.UploadID).Interface("resourceID", ev.ResourceID).Msg("cannot publish events - exiting")
+			return err
+		}
+	}
+
+	return nil
+}
+
+// process the scan
+func (av Antivirus) process(ev events.StartPostprocessingStep) (scanners.ScanResult, error) {
+	if ev.Filesize == 0 || (0 < av.m && av.m < ev.Filesize) {
+		return scanners.ScanResult{
+			Scantime: time.Now(),
+		}, nil
+	}
+
+	var err error
+	var rrc io.ReadCloser
+
+	switch ev.UploadID {
+	default:
+		rrc, err = av.downloadViaToken(ev.URL)
+	case "":
+		rrc, err = av.downloadViaReva(ev.URL, ev.Token, ev.RevaToken)
+	}
+	if err != nil {
+		av.l.Error().Err(err).Str("uploadid", ev.UploadID).Msg("error downloading file")
+		return scanners.ScanResult{}, err
+	}
+	defer rrc.Close()
+
+	res, err := av.s.Scan(rrc)
+	if err != nil {
+		av.l.Error().Err(err).Str("uploadid", ev.UploadID).Msg("error scanning file")
```

Comment:
we should return here, to prevent uploads when clamav isnt running. I see 
```
2023-03-22T17:19:30+01:00 ERR error scanning file error="dial unix /run/clamav/clamd.ctl: connect: no such file or directory" line=/home/jfd/Repositories/ocis/services/antivirus/pkg/service/service.go:163 service=antivirus uploadid=41f4b2b4-0b96-48e4-bbda-0a36b5acf8fe
2023-03-22T17:19:30+01:00 INF File scanned filename=logs_owncloud_ocis_20212_57_3.log infected=false line=/home/jfd/Repositories/ocis/services/antivirus/pkg/service/service.go:116 outcome=continue resourceID={"opaque_id":"4eb367b7-d870-45ec-acc2-cb7679c6b8e2","space_id":"f7fbf8c8-139b-4376-b307-cf0a8c2d0d9c"} service=antivirus uploadid=41f4b2b4-0b96-48e4-bbda-0a36b5acf8fe user=f7fbf8c8-139b-4376-b307-cf0a8c2d0d9c virus=
```
in the logs